### PR TITLE
Implementation of UI

### PR DESCRIPTION
**Pull request by thom776g**
fucky wucky

### CODE REVIEW

**Comment by thom776g** [comment]
Man kunne fjerne "Home screen" som er HVAC siden, da vi intet har med det - kode vis, men vi kunne også lade den blive, for at vise/sige at HVAC kontrollen/opkoblingen ville ske her - hvilket også stemmer nogenlunde overens med UI-design afsnittet

File: CentralHub.WebUI/Data/RoomService.cs [hunk 1]
```diff
@@ -32,6 +37,50 @@ public async Task RemoveRoomAsync(Room room, CancellationToken cancellationToken
...
+        request.Content = new StringContent(JsonSerializer.Serialize(addRoomRequest), Encoding.UTF8, "application/json");
+
+        var client = _clientFactory.CreateClient();
+
+        var response = await client.SendAsync(request, cancellationToken);
+
+        if (response.StatusCode != HttpStatusCode.OK)
+        {
+            throw new InvalidOperationException("Somethings fucky");
+        }
```
Comment (mads256h) [inline CentralHub.WebUI/Data/RoomService.cs:79]: Also check the AddRoomResponse message

File: CentralHub.WebUI/Data/TrackerService.cs [hunk 2]
```diff
@@ -13,6 +21,76 @@ public TrackerService(IHttpClientFactory clientFactory)
...
+            if (response.StatusCode != HttpStatusCode.OK)
+            {
+                throw new InvalidOperationException("Somethings fucky");
+            }
+        }
+        else
+        {
+            await RemoveTrackerAsync(tracker, cancellationToken);
+            UnregisteredTracker[] uTrackers = await GetUnregisteredTrackersAsync(cancellationToken);
+            UnregisteredTracker uTracker = uTrackers.Single(t => t.WifiMacAddress == tracker.WifiMacAddress);
```
Comment (mads256h) [inline CentralHub.WebUI/Data/TrackerService.cs:89]:
```suggestion
            var uTrackers = await GetUnregisteredTrackersAsync(cancellationToken);
            var uTracker = uTrackers.Single(t => t.WifiMacAddress == tracker.WifiMacAddress);
```

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 3]
```diff
@@ -1,96 +1,589 @@
...
+        public int TrackerCount { get; set; }
+        public bool IsExpanded { get; set; }
+    }
+
+    protected override async Task OnInitializedAsync()
+    {
+        _rooms = (await RoomService.GetRoomsAsync(_cancellationTokenSource.Token)).ToList();
+        _unregisteredTrackers = await TrackerService.GetUnregisteredTrackersAsync(_cancellationTokenSource.Token);
+
+        List<RoomWrapper> RoomWrappers = new List<RoomWrapper>();
```
Comment (mads256h) [inline CentralHub.WebUI/Pages/Rooms.razor:513]:
```suggestion
        var roomWrappers = new List<RoomWrapper>();
```

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 4]
```diff
@@ -1,96 +1,589 @@
...
+    }
     private async Task<int> GetTrackerCountAsync(Room room)
     {
         var trackers = await TrackerService.GetTrackersAsync(room, _cancellationTokenSource.Token);
 
         return trackers.Length;
     }
 
-    private async Task DeleteRoom(Room room)
+    protected async Task DeleteRoom(RoomWrapper wRoom)
```
Comment (mads256h) [inline CentralHub.WebUI/Pages/Rooms.razor:531]: Why is this protected and not private?

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 5]
```diff
@@ -1,96 +1,589 @@
...
         // Remove the room in memory
-        _trackerCountDictionary!.Remove(room);
-        _rooms!.Remove(room);
+        _rooms!.Remove(wRoom.Room);
+        _wRooms!.Remove(wRoom);
         
         StateHasChanged();
     }
 
+    protected async Task DeleteTracker(RoomWrapper wRoom, Tracker tracker)
```
Comment (mads256h) [inline CentralHub.WebUI/Pages/Rooms.razor:542]: private

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 6]
```diff
@@ -1,96 +1,589 @@
...
+        StateHasChanged();
+    }
+
     public void Dispose()
     {
         _cancellationTokenSource.Cancel();
         _cancellationTokenSource.Dispose();
     }
+
+void ToggleNode(RoomWrapper wRoom)
```
Comment (mads256h) [inline CentralHub.WebUI/Pages/Rooms.razor:559]: private

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 7]
```diff
@@ -1,96 +1,589 @@
...
+    {
+        wRoom.IsExpanded = !wRoom.IsExpanded;
+    }
+
+async Task HideModal(String modalId)
```
Comment (mads256h) [inline CentralHub.WebUI/Pages/Rooms.razor:564]: private

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 8]
```diff
@@ -1,96 +1,589 @@
...
+    {
+        await JSRuntime.InvokeVoidAsync("hideModal", modalId);
+        AssignStandardValues();
+        StateHasChanged();
+    }
+
+    async Task ShowModal(String modalId)
```
Comment (mads256h) [inline CentralHub.WebUI/Pages/Rooms.razor:571]: private

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 9]
```diff
@@ -1,96 +1,589 @@
...
+    {
+        await JSRuntime.InvokeVoidAsync("showModal", modalId);
+    }
+
+    void AssignStandardValues()
```
Comment (mads256h) [inline CentralHub.WebUI/Pages/Rooms.razor:576]: private

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 10]
```diff
@@ -1,96 +1,589 @@
...
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Room is null");
+        } 
+        else if(tracker == null)
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Tracker is null");
+        }
+    }
+
+    protected class RoomWrapper
```
Comment (mads256h) [inline CentralHub.WebUI/Pages/Rooms.razor:494]: private sealed

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 11]
```diff
@@ -1,96 +1,589 @@
...
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Room's ID is invalid");
+        } 
+        else if(name == "" || name == null)
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Please enter a name for the tracker");
+        }
+    }
+
+    async Task DeleteTrackerModal(RoomWrapper wRoom, Tracker tracker)
```
Comment (mads256h) [inline CentralHub.WebUI/Pages/Rooms.razor:476]: private

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 12]
```diff
@@ -1,96 +1,589 @@
...
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Room is null");
+        } 
+        else if(tracker == null)
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Tracker is null");
+        }
+    }
+
+    async Task ConfirmEditTracker(Tracker tracker, int roomId, String name, String description)
```
Comment (mads256h) [inline CentralHub.WebUI/Pages/Rooms.razor:453]: private

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 13]
```diff
@@ -1,96 +1,589 @@
...
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Please enter a name for the tracker");
+        } 
+        else if(trackerMacAdress == "-1")
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Tracker's MAC-Address is invalid");
+        }
+    }
+
+    async Task EditTrackerModal(RoomWrapper wRoom, Tracker tracker)
```
Comment (mads256h) [inline CentralHub.WebUI/Pages/Rooms.razor:431]: private

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 14]
```diff
@@ -1,96 +1,589 @@
...
+            await ShowModal("addTrackerModal");
+            StateHasChanged();
+        }
+        else
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Room is null");
+        }
+    }
 
+    async Task ConfirmAddTracker(int? roomId, String trackerMacAdress, String name, String description)
```
Comment (mads256h) [inline CentralHub.WebUI/Pages/Rooms.razor:407]: private

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 15]
```diff
@@ -1,96 +1,589 @@
...
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Room is null");
+        }
+        else if(name == "" || name == null)
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Please enter a name for the room");
+        }
+    }
+
+    async Task AddTrackerModal(RoomWrapper wRoom)
```
Comment (mads256h) [inline CentralHub.WebUI/Pages/Rooms.razor:390]: private

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 16]
```diff
@@ -1,96 +1,589 @@
...
+            await HideModal("editRoomModal");
+            StateHasChanged();
+        }
+        else
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Room is null");
+        }
     }
 
+    async Task ConfirmEditRoom(RoomWrapper wRoom, String name, String description)
```
Comment (mads256h) [inline CentralHub.WebUI/Pages/Rooms.razor:372]: private

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 17]
```diff
@@ -1,96 +1,589 @@
...
+        else
         {
-            var trackerCount = await GetTrackerCountAsync(room);
-            dictionary.Add(room, trackerCount);
+            await JSRuntime.InvokeVoidAsync("alert", "Room is null");
         }
+    }
 
-        _trackerCountDictionary = dictionary;
+    async Task DeleteRoomModal(RoomWrapper wRoom)
```
Comment (mads256h) [inline CentralHub.WebUI/Pages/Rooms.razor:358]: private

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 18]
```diff
@@ -1,96 +1,589 @@
...
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Please enter a name for the room");
+        }
+    }
 
-        var dictionary = new Dictionary<Room, int>();
-        
-        // FIXME: This is inefficient.
-        foreach (var room in _rooms)
+    async Task EditRoomModal(RoomWrapper wRoom)
```
Comment (mads256h) [inline CentralHub.WebUI/Pages/Rooms.razor:341]: private

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 19]
```diff
@@ -1,96 +1,589 @@
...
+    private String? _selectedTrackerMacAdress = "-1";
+    
+    async Task AddRoomModal()
+    {
+        await ShowModal("addRoomModal");
+        StateHasChanged();
+    }
 
-    protected override async Task OnInitializedAsync()
+    async Task ConfirmAddRoom(String name, String description)
```
Comment (mads256h) [inline CentralHub.WebUI/Pages/Rooms.razor:327]: private

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 20]
```diff
@@ -1,96 +1,589 @@
...
+    private RoomWrapper? _selectedWRoom;
+    private Tracker? _selectedTracker;
+    private int _selectedRoomId = -1;
+    private String _roomName = "Your room name"; 
+    private String _roomDescription = "Your room description";
+    private String _trackerName = "Your tracker name";
+    private String _trackerDescription = "Your tracker description"; 
+    private String? _selectedTrackerMacAdress = "-1";
+    
+    async Task AddRoomModal()
```
Comment (mads256h) [inline CentralHub.WebUI/Pages/Rooms.razor:321]: private

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 21]
```diff
@@ -1,96 +1,589 @@
...
+    private List<RoomWrapper> _wRooms = new List<RoomWrapper>();
+    private UnregisteredTracker[] _unregisteredTrackers = new UnregisteredTracker[0];
+    private RoomWrapper? _selectedWRoom;
+    private Tracker? _selectedTracker;
+    private int _selectedRoomId = -1;
+    private String _roomName = "Your room name"; 
+    private String _roomDescription = "Your room description";
+    private String _trackerName = "Your tracker name";
+    private String _trackerDescription = "Your tracker description"; 
+    private String? _selectedTrackerMacAdress = "-1";
```
Comment (mads256h) [inline CentralHub.WebUI/Pages/Rooms.razor:319]:
```suggestion
    private string _roomName = "Your room name"; 
    private string _roomDescription = "Your room description";
    private string _trackerName = "Your tracker name";
    private string _trackerDescription = "Your tracker description"; 
    private string? _selectedTrackerMacAddress = "-1";
```

**Review by mads256h — CHANGES_REQUESTED** [review]

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 22]
```diff
@@ -1,96 +1,587 @@
...
+        if(wRoom == null)
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Room is null");
+            return;
+        }
+
+        _selectedWRoom = wRoom;
+        _selectedRoomId = wRoom!.Room.RoomId;
+        _unregisteredTrackers = await TrackerService.GetUnregisteredTrackersAsync(_cancellationTokenSource.Token);
+        _selectedTrackerMacAddress = _unregisteredTrackers[0].WifiMacAddress;
```
Comment (Casper-NS) [inline CentralHub.WebUI/Pages/Rooms.razor:398]: This brokey, it throws an exception when there arent any unregistered trackers

**Review by mads256h — CHANGES_REQUESTED** [review]

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 23]
```diff
@@ -1,96 +1,587 @@
...
+    private async Task AddTrackerModal(RoomWrapper wRoom)
+    {
+        if(wRoom == null)
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Room is null");
+            return;
+        }
+
+        _selectedWRoom = wRoom;
+        _selectedRoomId = wRoom!.Room.RoomId;
```
Comment (Casper-NS) [inline CentralHub.WebUI/Pages/Rooms.razor:396]: Also no need for the exclamation mark when you have already checked for a null value.
This is just in general as there are several places where you do.

**Review by Casper-NS — CHANGES_REQUESTED** [review]
Good work, although a some fixes are nedded

File: CentralHub.WebUI/Pages/Rooms.razor (hunk 22 above)
Comment (thom776g) [reply CentralHub.WebUI/Pages/Rooms.razor]: fixed

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 24]
```diff
@@ -1,96 +1,587 @@
...
+        _unregisteredTrackers = await TrackerService.GetUnregisteredTrackersAsync(_cancellationTokenSource.Token);
+        _selectedTrackerMacAddress = _unregisteredTrackers[0].WifiMacAddress;
+        await ShowModal("addTrackerModal");
+        StateHasChanged();
+    }
+
+    private async Task ConfirmAddTracker(int? roomId, String trackerMacAdress, String name, String description)
```
[the remark this answers is not in the document]
Comment (thom776g) [reply CentralHub.WebUI/Pages/Rooms.razor:433]: Skal '?' ikke også fjernes fra "int? roomID" ???

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 25]
```diff
@@ -1,96 +1,587 @@
...
+        NavigationManager.NavigateTo(NavigationManager.Uri, forceLoad: true);
+    }
+
+    private async Task AddTrackerModal(RoomWrapper wRoom)
+    {
+        if(wRoom == null)
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Room is null");
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (thom776g) [reply CentralHub.WebUI/Pages/Rooms.razor:420]: Why not?

File: CentralHub.WebUI/Pages/Rooms.razor [hunk 26]
```diff
@@ -1,96 +1,587 @@
...
+        NavigationManager.NavigateTo(NavigationManager.Uri, forceLoad: true);
+    }
+
+    private async Task EditRoomModal(RoomWrapper wRoom)
+    {
+        if(wRoom == null)
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Room is null");
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (thom776g) [reply CentralHub.WebUI/Pages/Rooms.razor:373]: why not?

**Review by mads256h — APPROVED** [review]

**Review by Casper-NS — APPROVED** [review]

**Pull request merged**